asv_bench/benchmarks/rolling.py

```diff
@@ -0,0 +1,48 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import numpy as np
+import pandas as pd
+import xarray as xr
+
+from . import randn
+
+
+nx = 3000
+ny = 2000
+nt = 1000
+window = 20
+ds = xr.Dataset({'var1': (('x', 'y'), randn((nx, ny), frac_nan=0.1)),
+                 'var2': (('x', 't'), randn((nx, nt))),
+                 'var3': (('t', ), randn(nt))},
+                coords={'x': np.arange(nx),
+                        'y': np.linspace(0, 1, ny),
+                        't': pd.date_range('1970-01-01', periods=nt, freq='D'),
+                        'x_coords': ('x', np.linspace(1.1, 2.1, nx))})
+
+
+def time_rolling(func, center):
+    getattr(ds.rolling(x=window, center=center), func)()
+
+
+time_rolling.param_names = ['func', 'center']
+time_rolling.params = (['mean', 'count'], [True, False])
+
+
+def time_rolling_np(window_, min_periods):
+    ds.rolling(x=window_, center=False, min_periods=min_periods).reduce(
+        getattr(np, 'nanmean'))
+
+
+time_rolling_np.param_names = ['window_', 'min_periods']
+time_rolling_np.params = ([20, 40], [5, None])
```

Comment:
I expected the memory usage keeps almost the same (excluding the padded region) 
even if we changed the window size.
But with a doubled window size, the memory usage increases almost twice.

Is it the nature of numpy's reducing function? 
Or I'm doing something wrong...

Reply:
I've experienced something similar before. I think NumPy's reducing functions may need to copy data copy into a C or Fortran contiguous array to do computations.

Reply:
Probably the best way to get a definitive answer would be to ask on the NumPy issue tracker or mailing list.

Reply:
It looks that the rolling-reduction with a bare `np.ndarray` does not consume so much memory.
I think the problem is in my code...

Here is a script I checked with `np.ndarray`
```python
In [1]: import numpy as np
   ...: 
   ...: def rolling_window(a, window):
   ...:     shape = a.shape[:-1] + (a.shape[-1] - window + 1, window)
   ...:     strides = a.strides + (a.strides[-1],)
   ...:     return np.lib.stride_tricks.as_strided(a, shape=shape, strides=strid
   ...: es)
   ...: 
   ...: a = np.arange(20000000, dtype=float).reshape(1000, 20000)
   ...: a_rolling1 = rolling_window(a, 100)
   ...: a_rolling2 = rolling_window(a, 300)
   ...: a_rolling1.shape, a_rolling2.shape
Out[1]: ((1000, 19901, 100), (1000, 19701, 300))

In [2]: from memory_profiler import memory_usage
   ...: memory_usage((np.mean, (a_rolling1, ), {'axis': 2}))
Out[2]: -> 355 MB

In [3]: memory_usage((np.mean, (a_rolling2, ), {'axis': 2}))
Out[3]:  -> 354 MB
```

Reply:
OK. Found. `np.nanmean` copies the data.

Reply:
They seem to copy the original array with `nan` replaced by 0.
I think we can avoid this to replace `nan` *before* making the rolling window.